controller/volume_controller.go

```diff
@@ -888,10 +896,47 @@ func isFirstAttachment(rs map[string]*longhorn.Replica) bool {
 	return true
 }
 
+func isHealthyAndActiveReplica(r *longhorn.Replica) bool {
+	if r.Spec.FailedAt != "" {
+		return false
+	}
+	if r.Spec.HealthyAt == "" {
+		return false
+	}
+	if !r.Spec.Active {
+		return false
+	}
+	return true
+}
+
+func isDefinitelyHealthyAndActiveReplica(r *longhorn.Replica) bool {
+	if !isHealthyAndActiveReplica(r) {
+		return false
+	}
+	// An empty r.Spec.FailedAt doesn't necessarily indicate a healthy replica. The replica could be caught in an
+	// autosalvage loop. If it is, its r.Spec.FailedAt is repeatedly transitioning between empty and some time. Ensure
+	// the replica has become healthyAt since it last became failedAt.
+	// We know r.Spec.LastHealthyAt != "" because r.Spec.HealthyAt != "" from isHealthyAndActiveReplica.
+	if r.Spec.LastFailedAt != "" && !util.TimestampAfterTimestamp(r.Spec.LastHealthyAt, r.Spec.LastFailedAt) {
```

Comment:
You could save the ! by swapping the time arguments.  Unless you suspect that they might be equal, I suppose.  (But see my note below on argument order.)

Reply:
> This statement is right: An empty r.Spec.FailedAt doesn't necessarily indicate a healthy replica. But isHealthyAndActiveReplica(r) == true means r.Spec.HealthyAt != "" && r.Spec.FailedAt == "" ==> This should prove a replica is running/available.

I would like to agree, but I created this PR because this is NOT always the case. Sometimes, we have a replica with `r.Spec.HealthyAt != ""` that is completely corrupted. If it is selected for autosalvage, we clear `r.Spec.FailedAt` (`r.Spec.FailedAt == ""`). This fits our previous definition of "healthy", but we cannot rely on this definition when cleaning up stale replicas! If we do, we may clean up all other replicas in favor of a completely broken one.

> r.Spec.LastHealthyAt being earlier than r.Spec.LastFailedAt != "" ==> Replica is crashed and hasn't been successfully salvaged/reused.

Yes. In an "autosalvage loop", some replica is repeatedly crashed -> salvaged -> crashed -> salvaged -> crashed. If the replica has EVER failed (`r.Spec.LastFailedAt != ""`) and it hasn't "become healthy" (appeared as RW in an engine) since the last time it failed (`!util.TimestampAfterTimestamp(r.Spec.LastHealthyAt, r.Spec.LastFailedAt)`, we cannot rely on it. We must wait to clean up other replicas.

Reply:
I see. The salvaged replicas satisfy `r.Spec.HealthyAt != "" && r.Spec.FailedAt == ""`.  They may be invalid and can intervene in the subsequent replica cleanup executed.
And this function can filter out all failed replicas and salvaged but unverified replicas.

Reply:
> Sometimes, we have a replica with r.Spec.HealthyAt != "" that is completely corrupted. If it is selected for autosalvage, we clear r.Spec.FailedAt (r.Spec.FailedAt == ""). This fits our previous definition of "healthy", but we cannot rely on this definition when cleaning up stale replicas!

> The salvaged replicas satisfy r.Spec.HealthyAt != "" && r.Spec.FailedAt == "". They may be invalid and can intervene in the subsequent replica cleanup executed.
And this function can filter out all failed replicas and salvaged but unverified replicas.

Can you add more comments based on these discussions?

Reply:
I added an additional explanation to `isSafeAsLastReplica`. In addition, I put additional field explanations in the CRD.